Win32-network/src/System/Win32/Async/IOData.hsc

```diff
@@ -0,0 +1,366 @@
+             --
+             -- Note that the 'MVar' holds 'ErrorCode Async', even if
+             -- 'asyncType ~ WsaAsync', this is because the
+             -- 'GetQueuedCompletionStatus' and thus
+             -- 'dequeueCompletionPackets' can handle both `windows.h` and
+             -- `winsock2.h` apis.
+             --
+             -> IO a
+withIOCPData errorTag fd k = mask $ \unmask -> do
+    -- both 'IOData' and the stable pointer are allocated here and free either
+    --  * by 'IOManager' thread ('dequeueCompletionPackets')
+    --  * or here in case of some synchronous operations which do not push
+    --    a notifiction to IOCP.
+    -- We cannot simply use `allocaBytes` and make the allocation local, since
+    -- in case of async exception the IOCP thread will access already freed
+    -- memory.
+    --
+    -- TODO: hovewer, if we synchronously cancel operations we could use local
+    -- allocations for both 'IOData' and the stable pointer, which would make
+    -- this library much safer.  For that reason we woudl need to solve two problems:
+    --
+    --   *  'CancelIoEx' does not wait for an operation to be cancelled, we
+    --      would still need to do that.  The simplest solution which could work
+    --      is to await on the 'waitVar' (the thread that was waiting on it was
+    --      cancelled, so we could safely do that).
+    --
+    --   *  Find how to cancel asynchronous winsock2 operations.  Maybe
+    --   'CancelIoEx' works for them too?
+    --
+    ioDataPtr <- malloc
+    -- allocate stable pointer
+    (ioData, v) <- newIOData asyncTag
+    poke ioDataPtr ioData
+    result <-
+      (unmask $ k (iodOverlappedPtr asyncTag ioDataPtr) v)
+      `catch`
+         (\(e :: SomeAsyncException) -> do
+            -- CancelIoEx MSDN docs:
+            -- <https://docs.microsoft.com/en-us/windows/win32/fileio/cancelioex-func#remarks>
+            -- 
+            -- An MSDN example:
+            -- <https://docs.microsoft.com/en-us/windows/win32/fileio/canceling-pending-i-o-operations#canceling-asynchronous-io>
+            --
+            -- When we cancel an async operations an io completion packet the
+            -- io completion thread ('dequeueCompletionPackets') is notified.
+            -- 'GetQueuedCompletionStatus' sets error code to
+            -- 'ERROR_OPERATION_ABORTED';  thus the  'ioDataPtr' will be
+            -- freed by that thread.  There's no need for us to check the
+            -- status of the cancelation with 'GetOverlappedResult'
+            --
+            case fd of
+              FDHandle handle -> do
+                -- TODO: trace cancelResult and the error code
+                -- we need to cast 'Ptr (Overlapped Async)' to 'Ptr OVERLAPPED'.
+                void $ c_CancelIoEx handle (iodOverlappedPtr AsyncSing ioDataPtr)
+                throwIO e
+              FDSocket _ -> throwIO e
+                -- TODO: is there a way to cancel 'WSARecv'?
```

Comment:
This is important isn't it? These are the main ones that actually block, so the main ones we really want to cancel.